algosdk/testing/teal_blackbox.py

```diff
@@ -0,0 +1,720 @@
+            return predicate, lambda _: predicate
+
+        # N == 1:
+        return lambda args, actual: predicate(
+            args
+        ) == actual, lambda args: predicate(args)
+
+
+def mode_has_assertion(
+    mode: ExecutionMode, assertion_type: DryRunAssertionType
+) -> bool:
+    missing = {
+        ExecutionMode.Signature: {
+            DryRunAssertionType.cost,
+            DryRunAssertionType.lastLog,
+        },
+        ExecutionMode.Application: set(),
+    }
+    if assertion_type in missing[mode]:
+        return False
+
+    return True
+
+
+def dig_actual(
+    dryrun_resp: dict,
+    assert_type: DryRunAssertionType,
+    assertion_arg: Any = None,
+) -> Union[str, int, bool]:
+    txns = dryrun_resp["txns"]
+    assert (
+        len(txns) == 1
+    ), f"expecting exactly 1 transaction but got {len(txns)}"
+    txn = txns[0]
+    mode = (
+        ExecutionMode.Signature
+        if "logic-sig-messages" in txn
+        else ExecutionMode.Application
+    )
+    is_app = mode == ExecutionMode.Application
+
+    assert mode_has_assertion(
+        mode, assert_type
+    ), f"{mode} cannot handle dig information from txn for assertion type {assert_type}"
+    is_app = mode == ExecutionMode.Application
+
+    if assert_type == DryRunAssertionType.cost:
+        return txn["cost"]
+
+    if assert_type == DryRunAssertionType.lastLog:
+        last_log = txn.get("logs", [None])[-1]
+        if last_log is None:
+            return last_log
+        return b64decode(last_log).hex()
+
+    if assert_type == DryRunAssertionType.finalScratch:
+        trace = extract_trace(txn, is_app)
+        lines = extract_lines(txn, is_app)
+        bbr = scrape_the_black_box(trace, lines)
+        return {
+            k: v.as_python_type() for k, v in bbr.final_scratch_state.items()
+        }
+
+    if assert_type == DryRunAssertionType.stackTop:
+        trace = extract_trace(txn, is_app)
+        stack = trace[-1]["stack"]
+        if not stack:
+            return None
+        tv = TealVal.from_scratch(stack[-1])
+        return tv.as_python_type()
+
+    if assert_type == DryRunAssertionType.maxStackHeight:
+        return max(len(t["stack"]) for t in extract_trace(txn, is_app))
+
+    if assert_type == DryRunAssertionType.status:
+        return extract_status(mode, txn)
+
+    if assert_type == DryRunAssertionType.passed:
+        return extract_status(mode, txn) == "PASS"
+
+    if assert_type == DryRunAssertionType.rejected:
+        return extract_status(mode, txn) == "REJECT"
+
+    if assert_type == DryRunAssertionType.error:
+        ok, msg = assert_error(
+            dryrun_resp, pattern=assertion_arg, enforce=False
+        )
+        return ok or msg
+
+    if assert_type == DryRunAssertionType.noError:
+        ok, msg = assert_no_error(dryrun_resp, enforce=False)
+        return ok or msg
+
+    raise Exception(f"Unknown assert_type {assert_type}")
+
+
+def extract_logs(txn, decode_logs: bool = True):
```

Comment:
@tzaffi Minor:  It seems `decode_logs` is _not_ used within the method definition.  Is there logic worth adding or is it safe to remove the parameter?

Reply:
👍